app/models/mentor_management.rb

```diff
@@ -0,0 +1,122 @@
+  #   3.) If the team does not have a mentor.
+  def self.assign_mentor(assignment_id, team_id)
+    assignment = Assignment.find(assignment_id)
+    team = Team.find(team_id)
+
+    # RuboCop 'use guard clause instead of nested conditionals'
+    # return if assignments can't accept mentors
+    return unless assignment.auto_assign_mentor
+
+    # RuboCop 'use guard clause instead of nested conditionals'
+    # return if the assignment or team already have a topic
+    return if assignment.topics? || !team.topic.nil?
+
+    curr_team_size = Team.size(team_id)
+    max_team_members = Assignment.find(assignment_id).max_team_size
+
+    # RuboCop 'use guard clause instead of nested conditionals'
+    # return if the team size hasn't reached > 50% of capacity
+    return if curr_team_size * 2 <= max_team_members
+
+    # RuboCop 'use guard clause instead of nested conditionals'
+    # return if there's already a mentor in place
+    return if team.participants.any? { |participant| participant.duty == Participant::DUTY_MENTOR }
+
+    mentor_user = select_mentor(assignment_id)
+
+    # Add the mentor using team model class.
+    team_member_added = mentor_user.nil? ? false : team.add_member(mentor_user, assignment_id)
+
+    return unless team_member_added
+
+    notify_team_of_mentor_assignment(mentor_user, team)
+  end
+
+  def self.notify_team_of_mentor_assignment(mentor, team)
+    members = team.users
+    emails = members.map(&:email)
+    members_info = members.map{|mem| "#{mem.fullname} - #{mem.email}"}
+    mentor_info = "#{mentor.fullname} (#{mentor.email})"
+    message = "#{mentor_info} has been assigned as your mentor for assignment #{Assignment.find(team.parent_id).name} <br>Current members:<br> #{members_info.join('<br>')}"
+
+    Mailer.delayed_message(bcc: emails,
+                           subject: "[Expertiza]: New Mentor Assignment",
+                           body: message).deliver_now
+  end
+
+  # Returns true if [user] is a mentor, and false if not.
+  #
+  # [user] must be a User object.
+  #
+  # Checks the Participant relation to see if a row exists with
+  # user_id == user.id that also has 'mentor' in the duty attribute.
+  def self.user_a_mentor?(user)
+    Participant.exists?(user_id: user.id, duty: Participant::DUTY_MENTOR)
+  end
+
+  # Select all the participants who's duty in the participant
+  # table is [DUTY_MENTOR], and who are a participant of
+  # [assignment_id].
+  #
+  # @see participant.rb for the value of DUTY_MENTOR
+  def self.mentors_for_assignment(assignment_id)
+    Participant.where(parent_id: assignment_id, duty: Participant::DUTY_MENTOR)
+  end
+
+  # Produces a hash mapping mentor's user_ids to the aggregated
+  # number of teams they're part of, which acts as a proxy for
+  # the number of teams they're mentoring.
+  def self.zip_mentors_with_team_count(assignment_id)
+    mentor_ids = self.mentors_for_assignment(assignment_id).pluck(:user_id)
+
+    return [] if mentor_ids.empty?
```

Comment:
why return an array when the caller will be expecting a hash. potentially a method could be called on this array object when they expect a hash that could result in a nil:NilClass error

Reply:
The original idea was to return a hash, but `sort_by` returns an Array of Arrays and I forgot to update the method comment. I will fix this.

Reply:
sounds good.